swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,927 @@
+use crate::handler::{InboundUpgradeSend, OutboundUpgradeSend};
+use crate::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, IntoConnectionHandler,
+    KeepAlive, NegotiatedSubstream, NetworkBehaviourAction, NotifyHandler, SubstreamProtocol,
+};
+use futures::stream::{BoxStream, SelectAll};
+use futures::{Stream, StreamExt};
+use libp2p_core::connection::ConnectionId;
+use libp2p_core::either::EitherOutput;
+use libp2p_core::upgrade::{
+    DeniedUpgrade, EitherUpgrade, NegotiationError, ProtocolError, ReadyUpgrade,
+};
+use libp2p_core::{ConnectedPoint, PeerId, UpgradeError};
+use std::collections::{HashSet, VecDeque};
+use std::error::Error;
+use std::fmt;
+use std::future::Future;
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+use std::task::{Context, Poll, Waker};
+use std::time::{Duration, Instant};
+use void::Void;
+
+/// A low-level building block for protocols that can be expressed as async functions.
+///
+/// An async function in Rust is executed within an executor and thus only has access to its local state
+/// and by extension, the state that was available when the [`Future`] was constructed.
+///
+/// This [`ConnectionHandler`] aims to reduce the boilerplate for protocols which can be expressed as
+/// a static sequence of reads and writes to a socket where the response can be generated with limited
+/// "local" knowledge or in other words, the local state within the [`Future`].
+///
+/// For outbound substreams, arbitrary data can be supplied via [`InEvent::NewOutbound`] which will be
+/// made available to the callback once the stream is fully negotiated.
+///
+/// Inbound substreams may be opened at any time by the remote. To facilitate this one and more usecases,
+/// the supplied callbacks for inbound and outbound substream are given access to the handler's `state`
+/// field. State can be shared between the [`NetworkBehaviour`] and a [`FromFn`] [`ConnectionHandler`]
+/// via the [`Shared`] abstraction.
+///
+/// [`Shared`] tracks a piece of state and updates all registered [`ConnectionHandler`]s whenever the
+/// state changes.
+///
+/// [`NetworkBehaviour`]: crate::NetworkBehaviour
+pub fn from_fn(protocol: &'static str) -> Builder<WantState> {
+    Builder {
+        protocol,
+        phase: WantState {},
+    }
+}
+
+pub struct Builder<TPhase> {
+    protocol: &'static str,
+    phase: TPhase,
+}
+
+pub struct WantState {}
+
+pub struct WantInboundHandler<TState> {
+    state: Arc<TState>,
+}
+
+pub struct WantOutboundHandler<TState, TInbound> {
+    state: Arc<TState>,
+    max_inbound: usize,
+    inbound_handler: Option<
+        Box<
+            dyn Fn(
```

Comment:
AFAICS this restricts the part of the handler that constructs the stream processor to shared access to its closure — is there a reason for this? If not I’d prefer FnMut.

Reply:
No particular reason, I guess this could be useful if you want to track connection-local state across all inbound/outbound streams?

That would only be shared across one closure though, meaning you wouldn't be able to share state between the inbound and outbound factory function.

I think we might be able to change this interface to allow for a single factory function that gets called for both inbound and outbound streams which would allow you to share state. Is that what you would be after?

Reply:
Sharing state is not my concern because doing that always proceeds via shared references (of which Arc is a special case). `Fn` does not allow captured values to be modified (more precisely: mutably borrowed), even if they are owned.

Reply:
> `Fn` does not allow captured values to be modified (more precisely: mutably borrowed), even if they are owned.

If you want to mutate a capture value, you are effectively sharing the state within a connection across all stream handler factory invocations.

I think we are saying the same thing and I agree that this capability is useful :)

Reply:
Ah, you mean “sharing along the timeline”, not “sharing in different places”. Yes, now we both know that we are in agreement :-)